File: fastagi.pl
#!/usr/bin/env perl
#
# FastAGI Server example
#
# Copyright (C) 2014, Lefteris Zafiris <zaf.000@gmail.com>
#
# This program is free software, distributed under the terms of
# the GNU General Public License Version 2. See the LICENSE file
# at the top of the source tree.
#

package fastagi;

use strict;
use warnings;
use URI;
use Asterisk::AGI;
use base 'Net::Server::Fork';

use constant DEBUG => 0;

fastagi->run(
	{   proto       => 'tcp',
		port        => 4573,
		host        => '127.0.0.1',
		max_servers => 2000,
		log_level   => 0,
	}
);

sub process_request {
	my $self = shift;
	myagi();
}

sub myagi {
	my $status;
	my $file;
	my $agi   = new Asterisk::AGI;
	my %input = $agi->ReadParse();
	if (DEBUG) {
		warn "\n==AGI environment vars==\n";
		warn "$_: $input{$_}\n" foreach (keys %input);
	}
	my $uri   = URI->new($input{"request"});
	my %query = $uri->query_form;
	if (!exists $query{"file"}) {
		warn "No file name parameter passed, exiting.\n" if (DEBUG);
		goto HANGUP;
	}
	$file = $query{"file"};

	$status = $agi->channel_status('');
	if ($status != 6) {
		$status = $agi->answer('');
		if ($status == -1) {
			warn "Failed to answer channel\n";
			goto HANGUP;
		}
	}
	$agi->verbose("Playing back: $file", 0);
	$status = $agi->stream_file($file, '1234567890#*', 0);
	warn "Failed to playback file: $file\n" if ($status == -1);

HANGUP:
	$agi->hangup();
	return;
}

File: fastagi.go
// FastAGI server example in Go
//
// Copyright (C) 2014, Lefteris Zafiris <zaf@fastmail.com>
//
// This program is free software, distributed under the terms of
// the GNU General Public License Version 2. See the LICENSE file
// at the top of the source tree.

package main

import (
	"bufio"
	"log"
	"net"
	"net/url"

	"github.com/zaf/agi"
)

const debug = false

func main() {
	ln, err := net.Listen("tcp", ":4573")
	if err != nil {
		log.Fatal(err)
	}
	defer ln.Close()
	for {
		conn, err := ln.Accept()
		if err != nil {
			log.Println(err)
			continue
		}
		go connHandle(conn)
	}
}

func connHandle(c net.Conn) {
	defer c.Close()
	var rep agi.Reply
	var file string
	//Create a new FastAGI session
	myAgi := agi.New()
	rw := bufio.NewReadWriter(bufio.NewReader(c), bufio.NewWriter(c))
	err := myAgi.Init(rw)
	if err != nil {
		log.Printf("Error Parsing AGI environment: %v\n", err)
		return
	}
	if debug {
		// Print AGI environment variables.
		log.Println("AGI environment vars:")
		for key, value := range myAgi.Env {
			log.Printf("%-15s: %s\n", key, value)
		}
	}
	// Check passed arguments
	req, _ := url.Parse(myAgi.Env["request"])
	query, _ := url.ParseQuery(req.RawQuery)
	if query["file"] == nil {
		log.Println("No arguments passed, exiting...")
		goto HANGUP
	}
	file = query["file"][0]
	// Chech channel status and answer if not already answered
	rep, err = myAgi.ChannelStatus()
	if err != nil {
		log.Printf("AGI reply error: %v\n", err)
		return
	}
	if rep.Res != 6 {
		rep, err = myAgi.Answer()
		if err != nil || rep.Res == -1 {
			log.Printf("Failed to answer channel: %v\n", err)
			goto HANGUP
		}
	}
	// Playback file
	myAgi.Verbose("Playing back: "+file, 0)
	rep, err = myAgi.StreamFile(file, "1234567890#*")
	if err != nil {
		log.Printf("AGI reply error: %v\n", err)
		return
	}
	if rep.Res == -1 {
		log.Printf("Failed to playback file: %s\n", file)
	}
HANGUP:
	myAgi.Hangup()
	return
}
